### ISSUE TITLE
Possible race condition causing events to be dropped

### ISSUE THREAD

**Issue by olitomlinson** [Azure/azure-functions-durable-extension#573]
**Describe the bug**
I can't reliably reproduce this but I feel there is a race condition when awaiting on an activity, and then awaiting on an external event.

```cs
var submitToRouteResult = await context.CallActivityAsync<SubmitToRouteResult>("submit_to_route", @object);
 
// other orchestration code here (no activities)

var result = await context.WaitForExternalEvent<CoreDispatchAttemptResult>($"ChannelSubmissionResult_{route.Name}");
```

I have a scenario where the 'External Event' is being dropped due to the prior Activity not completing.

When I observe the History I see the following events in this order:

PartitionKey | RowKey | Timestamp | EventId | EventType | ExecutionId | IsPlayed | _Timestamp
-- | -- | -- | -- | -- | -- | -- | --
67d8b855-bd13-4cb5-b937-c330eb88b767 | 0000000000000000 | 2019-01-15T14:46:32.856Z | -1 | OrchestratorStarted | 0ccf5cd3f5ee4c84aaf93db34479e430 | false | 2019-01-15T14:46:32.640Z
67d8b855-bd13-4cb5-b937-c330eb88b767 | 0000000000000001 | 2019-01-15T14:46:32.857Z | -1 | ExecutionStarted | 0ccf5cd3f5ee4c84aaf93db34479e430 | true | 2019-01-15T14:41:16.686Z
67d8b855-bd13-4cb5-b937-c330eb88b767 | 0000000000000002 | 2019-01-15T14:46:32.857Z | 0 | TaskScheduled | 0ccf5cd3f5ee4c84aaf93db34479e430 | false | 2019-01-15T14:46:32.642Z
67d8b855-bd13-4cb5-b937-c330eb88b767 | 0000000000000003 | 2019-01-15T14:46:32.857Z | -1 | OrchestratorCompleted | 0ccf5cd3f5ee4c84aaf93db34479e430 | false | 2019-01-15T14:46:32.642Z
67d8b855-bd13-4cb5-b937-c330eb88b767 | 0000000000000004 | 2019-01-15T14:46:39.576Z | -1 | OrchestratorStarted | 0ccf5cd3f5ee4c84aaf93db34479e430 | false | 2019-01-15T14:46:39.551Z
67d8b855-bd13-4cb5-b937-c330eb88b767 | 0000000000000005 | 2019-01-15T14:46:39.576Z | -1 | EventRaised | 0ccf5cd3f5ee4c84aaf93db34479e430 | true | 2019-01-15T14:46:33.170Z
67d8b855-bd13-4cb5-b937-c330eb88b767 | 0000000000000006 | 2019-01-15T14:46:39.576Z | -1 | TaskCompleted | 0ccf5cd3f5ee4c84aaf93db34479e430 | true | 2019-01-15T14:46:33.240Z
67d8b855-bd13-4cb5-b937-c330eb88b767 | 0000000000000007 | 2019-01-15T14:46:39.576Z | -1 | OrchestratorCompleted | 0ccf5cd3f5ee4c84aaf93db34479e430 | false | 2019-01-15T14:46:39.552Z
67d8b855-bd13-4cb5-b937-c330eb88b767 | sentinel | 2019-01-15T14:46:39.577Z |   |   | 0ccf5cd3f5ee4c84aaf93db34479e430 |   |  

Notice above how '**EventRaised**' (Row Key 5) is being logged BEFORE '**TaskCompleted**' (RowKey 6)

This means my orchestrator waits for the event forever, but the event never comes as its already been dropped when it was actually waiting for the TaskComplete of the prior Activity Call.

**Other info**
1. Observing a different orchestration instance that completed successfully, the 'EventRaised' row is logged AFTER the 'TaskCompleted'

2. If I republish the external event (something that I can't do in a production scenario), the orchestration completes as expected.

**Investigative information**

- Durable Functions extension version: 1.7.0
- Function App version (1.0 or 2.0): 2.0
- Programming language used: c#

**Instance Id:** 
67d8b855-bd13-4cb5-b937-c330eb88b767
**Execution Id:**
0ccf5cd3f5ee4c84aaf93db34479e430
**Timestamp:**
2019-01-15T14:46:39.576Z
**region:**
UK West

**Comment by cgillum** [comment]
Thanks @olitomlinson, I think your analysis is spot-on.  There is a fundamental problem right now with the design of external events which doesn't allow it to work reliably in cases where external events are consumed at the same time as other tasks (in your case, the external event races with the activity function completion). More discussion of this general issue can be found here: https://github.com/Azure/azure-functions-durable-extension/issues/378

In the next feature release of Durable Functions (v1.8.0), I plan on introducing a new way to consume external events which will allow you to avoids this problem. Until then, the way to work around this is to avoid raising the external event until you know the orchestration is ready to consume it. One strategy is to use [custom orchestration status](https://docs.microsoft.com/en-us/azure/azure-functions/durable/durable-functions-custom-orchestration-status) to let the external event source know when it's safe to raise the event.

**Comment by cgillum** [comment]
Thanks for the details, I understand the constraints you're dealing with and I'm actually surprised that an issue/scenario like yours has not come up before, and it makes me realize that we need to deal with this more urgently.

In the meantime, is it possible to rearrange the API calls in your orchestration? For example, could you do something like this to mitigate the race?

```csharp
// subscribe to the async callback but don't yet await on it
var asyncCallbackReceivedTask = context.WaitForExternalEvent<CoreDispatchAttemptResult>(
    $"ChannelSubmissionResult_{route.Name}");

// schedule the activity which will result in the callback event above
var submitToRouteResult = await context.CallActivityAsync<SubmitToRouteResult>(
    "submit_to_route",
    @object);
 
// other orchestration code here (no activities)

// now wait for the callback to be received
var result = await asyncCallbackReceivedTask;
```

**Comment by olitomlinson** [comment]
This just blew my mind and I love it. 

Will roll it out in the morning and continue testing. Thanks so much. 

While one here, do you have much anecdotal evidence of DF adopters running many 100k of orchestrators (following the above code snippet) on Consumption Plan?

So far my out-of-the-box experience (on default host.json settings such as partition count 4) is yielding good* results on 10k-20k orchestrators being started as fast as possible(serviceBusTrigger). 

But when I exceed 20k orchestrators, the control queues start grinding to a halt, and nothing seems to progress. I tried 100k batch and had to drop the entire task hub after a few hours as it just was not doing much but a few control items every minute.

What’s weird is that it’s blistering fast for the first 10-20k and then drops off. This is where the functions host scale controller ‘black box’ concerns me... is it giving me lots of function host instances to clear the big ingestion service bus queue which feeds the orchestration starter? And then once all the work is *inside* DF, it’s not acknowledging the huge amount of work that needs processing on the control and work queues?

What I’m hypothesising is that the reason 10-20k flies through is it is riding on the coat-tails of a scaled out host due to the service bus ingestion queue, but after the ingestion queue clears, the scale controller fails to acknowledge the storage queue lengths inside DF?

**Comment by cgillum** [comment]
Cool - looking forward to hearing if this resolves your issue. In the meantime, we're working on a fix for this issue so that you won't need workarounds like this in the future.

I also moved your other question into a separate GitHub issue. :)

**Comment by olitomlinson** [comment]
Its going to be a hard one to test because I can't reliably recreate the race condition. 

I think I need to subscribe to the orchestrator life-cycle events over Event Grid and then observe how many complete events I receive in total and compare that to how many was expected. T

his may take me a few days to prove but I'll get back to you.

**Comment by cgillum** [comment]
Resolved in [v1.8.0](https://www.nuget.org/packages/Microsoft.Azure.WebJobs.Extensions.DurableTask/1.8.0) release.
